Userland/Utilities/pls.cpp

```diff
@@ -0,0 +1,213 @@
+        return 1;
+    }
+
+    // Unveil all paths in the user's PATH, as well as the command they've specified.
+    auto unveil_count = unveil_paths(command.at(0));
+    if (unveil_count == 0) {
+        warnln("Failed to unveil paths!");
+        return 1;
+    }
+
+    // Lock veil
+    unveil(nullptr, nullptr);
+
+    const char* username = getlogin();
+    auto sudoers_file_or_error = Core::File::open("/etc/sudoers", Core::IODevice::ReadOnly);
+    bool user_found = false;
+    if (sudoers_file_or_error.is_error()) {
+        warnln("couldn't open /etc/sudoers!");
+        return 1;
+    }
+
+    for (auto line = sudoers_file_or_error.value()->line_begin(); !line.at_end(); ++line) {
+        auto line_str = *line;
+
+        // Skip any comments
+        if (line_str.starts_with("#"))
+            continue;
+
+        // Our user is in the sudoers file!
+        if (line_str.to_string() == username) {
+            user_found = true;
+            break;
+        }
+    }
+
+    // User isn't in the sudoer's file
+    if (!user_found) {
+        warnln("{} is not in the sudoers file!", username);
+        return 2;
+    }
+
+    // The user was in the sudoers file, now let's ask for their password to ensure that it's actually them...
+    uid_t current_uid = getuid();
+    auto account_or_error = (username)
+        ? Core::Account::from_name(username)
+        : Core::Account::from_uid(current_uid);
+
+    if (account_or_error.is_error()) {
+        warnln("Core::Account::from_name: {}", account_or_error.error());
+        return 1;
+    }
+
+    const auto& account = account_or_error.value();
+    if (current_uid != 0 && account.has_password()) {
+        auto password = Core::get_password();
+        if (password.is_error()) {
+            warnln("{}", password.error());
+            return 1;
+        }
+
+        if (!account.authenticate(password.value().characters())) {
+            warnln("Incorrect or disabled password.");
+            return 1;
+        }
+    }
+
+    // TODO: Support swapping users instead of just defaulting to root
+    setgid(0);
+    setuid(0);
+
+    // Build the arguments list passed to `execvpe`
+    Vector<const char*> exec_args;
+    for (size_t i = 0; i < command.size(); i++) {
+        exec_args.append(command.at(i));
+    }
+
+    // Always terminate with a NULL (to signal end of args list)
+    exec_args.append(nullptr);
+
+    // Build the environment arguments
+    StringBuilder builder;
+
+    // Build SUDO_USER envvar
+    builder.append("SUDO_USER=");
+    builder.append(username);
+    auto sudo_user = builder.build();
+    builder.clear();
+
+    // Build SUDO_COMMAND envvar
+    builder.append("SUDO_COMMAND=");
+    builder.append(command.at(0));
+    auto sudo_command = builder.build();
+    builder.clear();
+
+    const char* envs[] = { "PROMPT=\\X\\u@\\h:\\w\\a\\e[33;1m\\h\\e[0m \\e[34;1m\\w\\e[0m \\p ",
+        "TERM=xterm", "PAGER=more", "PATH=/bin:/usr/bin:/usr/local/bin",
```

Comment:
It doesn't seem appropriate for this program to hard-code `PROMPT`, `TERM` and `PAGER`.
For `TERM`, we should probably just pass through whatever the current `TERM` is.
Do we need the other two at all?

Reply:
@awesomekling The `sudo user` and `sudo command` is a  feature that comes included with `sudo`. This way we can see who invoked `SUDO`. According to the Ubuntu user manual:

```
       If sudo is run by root and the SUDO_USER environment variable is set, the sudoers policy
       will use this value to determine who the actual user is.  This can be used by a user to
       log commands through sudo even when a root shell has been invoked.  It also allows the -e
       option to remain useful even when invoked via a sudo-run script or program.  Note,
       however, that the sudoers lookup is still done for root, not the user specified by
       SUDO_USER.
```